cs/src/Management/TunnelManagementClient.cs

```diff
@@ -644,936 +645,952 @@ private string UserLimitsPath
-                    case HttpStatusCode.TooManyRequests:
-                        throw new InvalidOperationException(errorMessage, hrex);
-
-                    case HttpStatusCode.Redirect:
-                    case HttpStatusCode.RedirectKeepVerb:
-                        // Add the redirect location to the exception data.
-                        // Normally the HTTP client should automatically follow redirects,
-                        // but this allows tests to validate the service's redirection behavior
-                        // when client auto redirection is disabled.
-                        hrex.Data["Location"] = response.Headers.Location;
-                        throw;
-
-                    default: throw;
-                }
-            }
-
-            throw new Exception(errorMessage, innerException);
-        }
-
-        /// <summary>
-        /// Error details that may be returned from the service with 500 status responses
-        /// (when in development mode).
-        /// </summary>
-        /// <remarks>
-        /// Copied from Microsoft.VsSaaS.Common to avoid taking a dependency on that assembly.
-        /// </remarks>
+                            }
+                                
+                            if (problemDetails.Errors != null)
+                            {
+                                foreach (var error in problemDetails.Errors)
+                                {
+                                    var messages = string.Join(" ", error.Value);
+                                    errorMessage += $"\n{error.Key}: {messages}";
+                                }
+                            }
+                        }
+                    }
+                    else if ((int)response.StatusCode >= 500)
+                    {
+                        // 5xx status responses may include VS SaaS error details.
+                        var errorDetails = await response.Content.ReadFromJsonAsync<ErrorDetails>(
+                            JsonOptions, cancellation);
+                        if (!string.IsNullOrEmpty(errorDetails?.Message))
+                        {
+                            errorMessage = "Tunnel service error: " + errorDetails!.Message;
+                            if (!string.IsNullOrEmpty(errorDetails.StackTrace))
+                            {
+                                errorMessage += "\n" + errorDetails.StackTrace;
+                            }
+                        }
+                    }
+                }
+                catch (Exception ex)
+                {
+                    // A default error message will be filled in below.
+                    innerException = ex;
+                }
+            }
+
+            if (errorMessage == null &&
+                (int)response.StatusCode >= 400 && (int)response.StatusCode < 500 &&
+                (!response.Headers.TryGetValues("X-Served-By", out var servedBy) ||

```

Comment:
Just for my own understanding, what does the X-Served-By header have to do with detecting a firewall? Is it just assumed if the returned status code is a 4xx and that this header values indicates it couldn't reach the service then we can assume that it was due to a firewall?

Reply:
The tunnel service always includes the header `X-Served-By: tunnels-cluster-....` in every response. If we get a response that doesn't have that header, or has a different value), that means the response was not from the tunnel service. So the only other possibility is that the response was from a proxy or firewall that intercepted and blocked the request.